llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,371 @@
+//
+// Implementation of a Loop Register rewriting pass, that reduces WAW physical
+// register conflicts, so that software pipelining can operate on loops.
+//
+//===----------------------------------------------------------------------===//
+
+#include "AIEBaseInstrInfo.h"
+#include "AIEBaseRegisterInfo.h"
+#include "Utils/AIELoopUtils.h"
+
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/CodeGen/LiveDebugVariables.h"
+#include "llvm/CodeGen/LiveIntervals.h"
+#include "llvm/CodeGen/LiveRegMatrix.h"
+#include "llvm/CodeGen/LiveStacks.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstr.h"
+#include "llvm/CodeGen/MachineLoopInfo.h"
+#include "llvm/CodeGen/MachineOperand.h"
+#include "llvm/CodeGen/Passes.h"
+#include "llvm/CodeGen/SlotIndexes.h"
+#include "llvm/CodeGen/TargetInstrInfo.h"
+#include "llvm/CodeGen/TargetSubtargetInfo.h"
+#include "llvm/MC/MCRegister.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+#include <llvm/CodeGen/MachineBasicBlock.h>
+#include <map>
+
+using namespace llvm;
+
+#define DEBUG_TYPE "aie-waw-reg-rewrite"
+
+namespace {
+
+///
+/// This rewrites Registers, if they are redefined in the same Loop body and
+/// could cause WAW issues in the loop pipelining.
+/// The pass will update the \p VirtRegMap so that the new vregs have fixed
+/// assignments, guaranteeing that the reg renaming does not cause conflicts.
+class AIEWawRegRewriter : public MachineFunctionPass {
+
+public:
+  static char ID;
+  AIEWawRegRewriter() : MachineFunctionPass(ID) {}
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesCFG();
+    AU.addRequired<VirtRegMap>();
+    AU.addPreserved<VirtRegMap>();
+    // no MachineInstr are added, therefore the SlotIndexes remain constant and
+    // preserved
+    AU.addRequired<SlotIndexes>();
+    AU.addPreserved<SlotIndexes>();
+    // no new Virtual Register are generated, therefore the LiveDebugVariables
+    // do not have to be updated
+    AU.addRequired<LiveDebugVariables>();
+    AU.addPreserved<LiveDebugVariables>();
+    AU.addRequired<LiveStacks>();
+    AU.addPreserved<LiveStacks>();
+    AU.addRequired<LiveIntervals>();
+    AU.addPreserved<LiveIntervals>();
+    AU.addRequired<LiveRegMatrix>();
+    AU.addPreserved<LiveRegMatrix>();
+    AU.addRequired<MachineLoopInfo>();
+    AU.addPreserved<MachineLoopInfo>();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+
+  bool runOnMachineFunction(MachineFunction &Fn) override;
+
+private:
+  MachineFunction *MF = nullptr;
+  const MachineRegisterInfo *MRI = nullptr;
+  const AIEBaseRegisterInfo *TRI = nullptr;
+  VirtRegMap *VRM = nullptr;
+  LiveRegMatrix *LRM = nullptr;
+  LiveIntervals *LIS = nullptr;
+  bool Modified = false;
+
+  bool renameMBBPhysRegs(const MachineBasicBlock *MBBs);
+  SmallSet<MCPhysReg, 16>
+  getMBBForbiddenReplacements(const MachineBasicBlock *MBB);
```

Comment:
Could you add a docstring for getMBBForbiddenReplacements? I will forget what is considered "forbidden" in a couple of weeks :)

Reply:
added a comment to clarify its intention